File: express/03-avanzado/01-api-testing/src/routes/productos.js
const express = require('express')
const router = express.Router()
const { 
  getProductos, 
  getProductoById, 
  crearProducto, 
  actualizarProducto, 
  eliminarProducto,
  getCategorias 
} = require('../data/productos')

router.get('/', (req, res) => {
  const { categoria, disponible, minPrecio, maxPrecio } = req.query
  
  const filtros = {
    categoria,
    disponible: disponible !== undefined ? disponible === 'true' : undefined,
    minPrecio: minPrecio ? parseFloat(minPrecio) : undefined,
    maxPrecio: maxPrecio ? parseFloat(maxPrecio) : undefined
  }
  
  const productos = getProductos(filtros)
  res.json({ success: true, data: productos })
})

router.get('/categorias', (req, res) => {
  res.json({ success: true, data: getCategorias() })
})

router.get('/:id', (req, res) => {
  const id = parseInt(req.params.id)
  const producto = getProductoById(id)
  
  if (!producto) {
    return res.status(404).json({ success: false, message: 'Producto no encontrado' })
  }
  
  res.json({ success: true, data: producto })
})

router.post('/', (req, res) => {
  const { nombre, precio, categoria, descripcion, disponible } = req.body
  
  if (!nombre || !precio) {
    return res.status(400).json({ success: false, message: 'Nombre y precio requeridos' })
  }
  
  if (precio < 0) {
    return res.status(400).json({ success: false, message: 'El precio no puede ser negativo' })
  }
  
  const producto = crearProducto({ 
    nombre, 
    precio: parseFloat(precio), 
    categoria: categoria || 'General',
    descripcion: descripcion || '',
    disponible: disponible !== false
  })
  
  res.status(201).json({ success: true, data: producto })
})

router.put('/:id', (req, res) => {
  const id = parseInt(req.params.id)
  const producto = actualizarProducto(id, req.body)
  
  if (!producto) {
    return res.status(404).json({ success: false, message: 'Producto no encontrado' })
  }
  
  res.json({ success: true, data: producto })
})

router.delete('/:id', (req, res) => {
  const id = parseInt(req.params.id)
  const eliminado = eliminarProducto(id)
  
  if (!eliminado) {
    return res.status(404).json({ success: false, message: 'Producto no encontrado' })
  }
  
  res.json({ success: true, message: 'Producto eliminado' })
})

module.exports = router

File: fastapi/02-intermedio/02-productos/src/routes/productos.py
from fastapi import APIRouter, HTTPException
from typing import Optional
from .schemas.producto import ProductoCreate, ProductoUpdate
from .repositories.producto import producto_repo

router = APIRouter(prefix="/productos", tags=["productos"])


@router.get("", response_model=dict)
def listar_productos(
    categoria: Optional[str] = None,
    disponible: Optional[bool] = None,
    min_precio: Optional[float] = None,
    max_precio: Optional[float] = None,
    busqueda: Optional[str] = None
):
    productos = producto_repo.get_all(categoria, disponible, min_precio, max_precio, busqueda)
    return {"success": True, "data": productos}


@router.get("/{id}", response_model=dict)
def obtener_producto(id: int):
    producto = producto_repo.get_by_id(id)
    if not producto:
        raise HTTPException(status_code=404, detail="Producto no encontrado")
    return {"success": True, "data": producto}


@router.post("", response_model=dict, status_code=201)
def crear_producto(producto: ProductoCreate):
    nuevo_producto = producto_repo.create(producto)
    return {"success": True, "data": nuevo_producto}


@router.put("/{id}", response_model=dict)
def actualizar_producto(id: int, producto: ProductoUpdate):
    producto_actualizado = producto_repo.update(id, producto)
    if not producto_actualizado:
        raise HTTPException(status_code=404, detail="Producto no encontrado")
    return {"success": True, "data": producto_actualizado}


@router.delete("/{id}", response_model=dict)
def eliminar_producto(id: int):
    if not producto_repo.delete(id):
        raise HTTPException(status_code=404, detail="Producto no encontrado")
    return {"success": True, "message": "Producto eliminado"}


@router.post("/{id}/stock", response_model=dict)
def actualizar_stock(id: int, cantidad: int):
    try:
        producto = producto_repo.update_stock(id, cantidad)
        if not producto:
            raise HTTPException(status_code=404, detail="Producto no encontrado")
        return {"success": True, "data": producto}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))